From 2ac5d27aad168598a2504f0ee43952a7043916ec Mon Sep 17 00:00:00 2001
From: "hjnim.kim" <hjnim.kim@samsung.com>
Date: Fri, 8 Mar 2013 11:24:06 +0900
Subject: [PATCH] JQM Apply tizen style popup

Apply tizen style popup ( data-position-to null or origin )

Change-Id: I53553b7fda18f64d7b696fdc8a4eb695046a8db0
---
 libs/js/jquery-mobile-1.2.0/js/widgets/popup.js |   98 ++++++++++++++++++++---
 1 files changed, 88 insertions(+), 10 deletions(-)

diff --git a/libs/js/jquery-mobile-1.2.0/js/widgets/popup.js b/libs/js/jquery-mobile-1.2.0/js/widgets/popup.js
index 8174055..aa71dbc 100644
--- a/libs/js/jquery-mobile-1.2.0/js/widgets/popup.js
+++ b/libs/js/jquery-mobile-1.2.0/js/widgets/popup.js
@@ -58,7 +58,7 @@ define( [ "jquery",
 			//      https://github.com/jquery/jquery-mobile/issues/4784
 			//
 			// NOTE this option is modified in _create!
-			history: !$.mobile.browser.ie
+			history: false
 		},
 
 		_eatEventAndClose: function( e ) {
@@ -109,7 +109,7 @@ define( [ "jquery",
 		},
 
 		_resizeTimeout: function() {
-			if ( !this._maybeRefreshTimeout() ) {
+			if ( !this._maybeRefreshTimeout() && this.positionTo === "window" ) {
 				// effectively rapid-open the popup while leaving the screen intact
 				this._trigger( "beforeposition" );
 				this._ui.container
@@ -144,7 +144,8 @@ define( [ "jquery",
 			var ui = {
 					screen: $( "<div class='ui-screen-hidden ui-popup-screen'></div>" ),
 					placeholder: $( "<div style='display: none;'><!-- placeholder --></div>" ),
-					container: $( "<div class='ui-popup-container ui-selectmenu-hidden'></div>" )
+					container: $( "<div class='ui-popup-container ui-selectmenu-hidden'></div>" ),
+					arrow : $("<div class='ui-arrow'></div>")
 				},
 				thisPage = this.element.closest( ".ui-page" ),
 				myId = this.element.attr( "id" ),
@@ -174,7 +175,7 @@ define( [ "jquery",
 				ui.placeholder.html( "<!-- placeholder for " + myId + " -->" );
 			}
 			ui.container.append( this.element );
-
+			ui.container.append( ui.arrow );
 			// Add class to popup element
 			this.element.addClass( "ui-popup" );
 
@@ -276,7 +277,7 @@ define( [ "jquery",
 		},
 
 		_setTolerance: function( value ) {
-			var tol = { t: 30, r: 15, b: 30, l: 15 };
+			var tol = { t: 5, r: 5, b: 5, l: 5 };
 
 			if ( value ) {
 				var ar = String( value ).split( "," );
@@ -361,7 +362,11 @@ define( [ "jquery",
 					cx: winCoords.cx - this._tolerance.l - this._tolerance.r,
 					cy: winCoords.cy - this._tolerance.t - this._tolerance.b
 				},
-				menuSize, ret;
+				menuSize, ret,
+				linkOffset = $(this.link).offset(),
+				positionOffsets = [],
+				correctionValue = [0,0],
+				arrayIdx;
 
 			// Clamp the width of the menu before grabbing its size
 			this._ui.container.css( "max-width", rc.cx );
@@ -389,7 +394,48 @@ define( [ "jquery",
 
 			ret.y -= Math.min( ret.y, Math.max( 0, ret.y + menuSize.cy - docHeight ) );
 
-			return { left: ret.x, top: ret.y };
+			if ( this.positionTo !== "origin" )
+			{
+				return { left: ret.x, top: ret.y , arrowleft: 0 , arrowtop: 0};
+			}
+
+			positionOffsets = [ linkOffset.left,
+								linkOffset.top,
+								docEl.clientHeight - ( linkOffset.top + $(this.link).height() ),
+								docEl.clientWidth - ( linkOffset.left + $(this.link).width() )];
+			arrayIdx = positionOffsets.indexOf(Math.max.apply(window,positionOffsets));
+
+			switch( arrayIdx )
+			{
+				case 0:
+					correctionValue = [ -$(this.link).width() , 0];
+					arrow.attr( "class", "" )
+									.addClass( "ui-arrow right" );
+					break;
+				case 1:
+					correctionValue = [ 0 , -(ret.y + menuSize.cy - linkOffset.top)];
+					arrowtop = menuSize.cy - 2;
+					arrowleft = (linkOffset.left - ret.x + correctionValue[0]) + ( $(this.link).width() / 2 ) - parseInt( $(this._ui.arrow).css("border-width") ) / 2;
+					$(this._ui.arrow).attr( "class", "" )
+									.addClass( "ui-arrow bottom" );
+					break;
+				case 2:
+					correctionValue = [ 0 , ( linkOffset.top + $(this.link).height() - ret.y ) ];
+					arrowtop = - parseInt( $(this._ui.arrow).css("border-width") ) * 2 + 1;
+					arrowleft = (linkOffset.left - ret.x + correctionValue[0]) + ( $(this.link).width() / 2 ) - parseInt( $(this._ui.arrow).css("border-width") ) / 2;
+					$(this._ui.arrow).attr( "class", "" )
+									.addClass("ui-arrow top");
+					break;
+				case 3:
+					correctionValue = [ ( menuSize.cx < $(this.link).width() ) ? ( $(this.link).width() / 2 ) + ( menuSize.cx / 2) : $(this.link).width() , 0];
+					arrowtop = ( linkOffset.top - ret.y  ) + ( $(this.link).height() / 2 ) - parseInt( $(this._ui.arrow).css("border-width") ) ;
+					arrowleft = - parseInt( $(this._ui.arrow).css("border-width") ) * 2;
+					$(this._ui.arrow).attr( "class", "" )
+									.addClass("ui-arrow right");
+					break;
+			}
+
+			return { left: ret.x + correctionValue[0], top: ret.y + correctionValue[1] , arrowleft: arrowleft , arrowtop: arrowtop };
 		},
 
 		_createPrereqs: function( screenPrereq, containerPrereq, whenDone ) {
@@ -569,7 +615,7 @@ define( [ "jquery",
 			this._ui.container
 				.removeClass( "ui-selectmenu-hidden" )
 				.offset( coords );
-
+			this._ui.arrow.css( { top : coords.arrowtop, left : coords.arrowleft } );
 			if ( this.options.overlayTheme && androidBlacklist ) {
 				/* TODO:
 				The native browser on Android 4.0.X ("Ice Cream Sandwich") suffers from an issue where the popup overlay appears to be z-indexed
@@ -694,15 +740,47 @@ define( [ "jquery",
 		// what should be in _open. Seems to be "visual" vs "history" for now
 		open: function( options ) {
 			var self = this, opts = this.options, url, hashkey, activePage, currentIsDialog, hasHash, urlHistory;
-
+			// self.link = ( $(event.target).attr('data-role') === 'button') ? event.target : $(event.target).closest('[data-role="button"]')[0];
 			// make sure open is idempotent
 			if( $.mobile.popup.active ) {
 				return;
 			}
-
 			// set the global popup mutex
 			$.mobile.popup.active = this;
+			if( !options ) {
+				options = [];
+			}
+			if ( !options.link ) {
+				self.link = ( $(event.target).closest('a')[0] || $(event.target).closest('div')[0] );
+			} else {
+				self.link = options.link;
+			}
+			self.positionTo = ( options != null && options.positionTo != null ) ? options.positionTo : "origin";
+			if ( $(self.link).jqmData("position-to") !== "window"
+					&& self.positionTo !== "window" ) {
+
+				$(self.element).addClass("ui-ctxpopup");
+				$(self._ui.container).removeClass("ui-popup-container")
+					.addClass("ui-ctxpopup-container");
 
+				if( self.positionTo !== "origin" ) {
+					$(self._ui.arrow).hide();
+				} else {
+					$(self._ui.arrow).show();
+				}
+			} else {
+				$(self._ui.arrow).hide();
+				// apply opacity back screen
+				this._setOverlayTheme( "dim" );
+			}
+			if( !options.x
+				&& self.positionTo === "origin" ) {
+				options.x = $(self.link).offset().left + $(self.link).outerWidth() / 2;
+			}
+			if( !options.y
+				&& self.positionTo === "origin" ) {
+				options.y = $(self.link).offset().top + $(self.link).outerHeight() / 2;
+			}
 			// if history alteration is disabled close on navigate events
 			// and leave the url as is
 			if( !( opts.history ) ) {
-- 
1.7.5.4

